rustls/src/client/hs.rs

```diff
@@ -271,6 +272,39 @@ fn emit_client_hello_for_retry(
     // Do we have a SessionID or ticket cached for this host?
     let tls13_session = prepare_resumption(&input.resuming, &mut exts, suite, cx, config);
 
+    // Extensions MAY be randomized
+    // but they also need to keep the same order as the the previous ClientHello
+    exts.sort_by_cached_key(|new_ext| {
+        // PSK extension is always last
+        if let ClientExtension::PresharedKey(..) = new_ext {
+            return u32::MAX;
+        }
+
+        let seed = input.hello.extension_order_seed | (new_ext.ext_type().get_u16() as u32);
+        let key = low_quality_integer_hash(seed);
+        if key == u32::MAX {
+            0
+        } else {
+            key
+        }
+    });
+
+    fn low_quality_integer_hash(mut x: u32) -> u32 {
+        x = x
+            .wrapping_add(0x7ed55d16)
+            .wrapping_add(x << 12);
+        x = (x ^ 0xc761c23c) ^ (x >> 19);
+        x = x
+            .wrapping_add(0x165667b1)
+            .wrapping_add(x << 5);
+        x = x.wrapping_add(0xd3a2646c) ^ (x << 9);
+        x = x
+            .wrapping_add(0xfd7046c5)
+            .wrapping_add(x << 3);
+        x = (x ^ 0xb55a4f09) ^ (x >> 16);
+        x
+    }
```

Comment:
I think I'd prefer to move this out of the way, to the bottom of the module.

Not sure if the other maintainers agree?